lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,983 @@
-};
+/// A namespace that is used to store existing names and generate names.  This
+/// exists to work around limitations of SymbolTables.
+class CircuitNamespace {
+  StringSet<> internal;
 
-/// Convert an arbitrary attributes into an optional AugmentedField.  Returns
-/// None if the attribute is an invalid AugmentedField.
-static Optional<AugmentedField> decodeField(Attribute maybeField) {
-  auto field = maybeField.dyn_cast_or_null<DictionaryAttr>();
-  if (!field)
-    return {};
-  auto tpeString = field.getAs<StringAttr>("tpe");
-  auto name = field.getAs<StringAttr>("name");
-  if (!name || !tpeString)
-    return {};
-  auto tpe = llvm::StringSwitch<ElementInfo::Kind>(tpeString.getValue())
-                 .Case("sifive.enterprise.grandcentral.AugmentedBundleType",
-                       ElementInfo::Bundle)
-                 .Case("sifive.enterprise.grandcentral.AugmentedVectorType",
-                       ElementInfo::Vector)
-                 .Case("sifive.enterprise.grandcentral.AugmentedGroundType",
-                       ElementInfo::Ground)
-                 .Case("sifive.enterprise.grandcentral.AugmentedStringType",
-                       ElementInfo::String)
-                 .Case("sifive.enterprise.grandcentral.AugmentedBooleanType",
-                       ElementInfo::Boolean)
-                 .Case("sifive.enterprise.grandcentral.AugmentedIntegerType",
-                       ElementInfo::Integer)
-                 .Case("sifive.enterprise.grandcentral.AugmentedDoubleType",
-                       ElementInfo::Double)
-                 .Default(ElementInfo::Error);
-  if (tpe == ElementInfo::Error)
-    return {};
-
-  StringRef description = {};
-  if (auto maybeDescription = field.getAs<StringAttr>("description"))
-    description = maybeDescription.getValue();
-  return Optional<AugmentedField>({name.getValue(), description, tpe});
-}
-
-/// Convert an Annotation into an optional AugmentedBundleType.  Returns None if
-/// the annotation is not an AugmentedBundleType.
-static Optional<AugmentedBundleType> decodeBundleType(Annotation anno) {
-  auto defName = anno.getMember<StringAttr>("defName");
-  auto elements = anno.getMember<ArrayAttr>("elements");
-  if (!defName || !elements)
-    return {};
-  AugmentedBundleType bundle(
-      {defName.getValue(), SmallVector<AugmentedField>()});
-  for (auto element : elements) {
-    auto field = decodeField(element);
-    if (!field)
-      return {};
-    bundle.elements.push_back(field.getValue());
+public:
+  /// Construct a new namesapce from a circuit op.  This namespace will be
+  /// composed of any operation in the first level of the circuit that contains
+  /// a symbol.
+  CircuitNamespace(Operation *op) {
+    op->walk([&](Operation *op) {
+      if (auto symbol =
+              op->getAttrOfType<StringAttr>(SymbolTable::getSymbolAttrName()))
+        internal.insert(symbol.getValue());
+      return WalkResult::skip();
+    });
```

Comment:
I don't think it makes sense to use `walk` when you never recurse:
 ```c++
  CircuitNamespace(CircuitOp circuit) {
    for (auto *op : circuit.getBodyBlock())
      if (auto symbol =
        op->getAttrOfType<StringAttr>(SymbolTable::getSymbolAttrName()))
          internal.insert(symbol.getValue());
  }
```
 Or maybe you can use `FModuleLIke`
```c++
for (auto ops : circuit.getBodyBlock().getOps<FModuleLike>()) 
  internal.insert(op.moduleName());
```

Reply:
I'll switch to the first one. I'm intentionally not trying to match on `FModuleLike` because I may have SV dialect stuff whose symbols I want to include.